### DISCUSSION TITLE
relative positioning of objects

### DISCUSSION THREAD

**Discussion by tom-hoffman** [libfive/libfive#527] [Q&A] [answered]
I'm teaching a 9th grade intro to computer science course, and I'm working on ideas for using some 3d printers I ended up with.  Our school in Providence is literally in a converted jewelry factory, so I'm thinking about having kids design beads and baubles, which seems like it could be sufficiently straightforward for them.

What drew me to looking at libfive is that in the past I have used the [Bootstrap](https://bootstrapworld.org/) introductory curriculum using their Racket/Scheme image library.  So in particular I'm wondering how hard it would be (for me, not you) to implement a version of their [API for overlaying images](https://docs.racket-lang.org/teachpack/2htdpimage.html#%28part._.Overlaying_.Images%29).  That would let me teach a nice sequence from 2d to 3d exercises.

The basic thing I don't see in your documentation is whether I can get a shape to return information about its size and position after it is created, to easily make a function that would, say, put one image "above" another.  

_edit -> a bit later..._ I'm realizing I need to think a bit more deeply about what this 2d library actually does (it has been a few years).  It actually lets you make basic shapes without giving them a position in the coordinate plane, and then it lets you position shapes relative to each other since the functions are not changing the images but returning new ones.  Sorry, I've been teaching basic graphics with Javascript lately and am a little out of the functional mode.  So I don't have a specific question at this point... (but don't seem to be able to delete the topic).

At this point I've just dipped my toe into libfive, so just a nudge in the right direction is what I'm looking for.  I know some Racketeers at Brown who may be able to help me further if it seems like this would work.

Thanks!

**Comment by mkeeter** [comment, marked as answer]
It sounds like you've mostly answered your own question!  In the same way that Racket's functions operate on pixels, `libfive` doesn't know much about shapes after they're called.  For example, calling `sphere` will give you a handle to a shape, but the fact that it's a sphere is forgotten; it's simply a black-box that **when evaluated** happens to be a sphere.

You can then combine it with other shapes using `union` / `intersection` / `difference`, in the same way that Racket has `overlay` to combine images; each of these functions returns a new black-box handle, in the same way that the Racket functions return an array of pixels (without, for example, details on the CSG tree that built them).

Right now, there's no support for colors, which could be a sticking point for the analogy.

Hope that helps!

**Comment by tom-hoffman** [comment]
Here's an interesting comparison between the two libraries:

2d -- [HtDP/2e Racket](https://docs.racket-lang.org/teachpack/2htdpimage-guide.html#%28part._.Recursive_.Image_.Functions%29)

```
(define (sierpinski-carpet n)
  (cond
    [(zero? n) (square 1 "solid" "black")]
    [else
      (local [(define c (sierpinski-carpet (- n 1)))
              (define i (square (image-width c) "solid" "white"))]
        (above (beside c c c)
               (beside c i c)
               (beside c c c)))]))
```
3d -- menger sponge from libfive examples:

```
(define (recurse center scale i)
  (define base
    (rectangle (- center scale) (+ center scale)))
  (if (= i 0)
    base
    (apply union (cons base
      (map (lambda (o)
        (recurse (+ center (* 2 o scale))
                 (/ scale 3)
                 (1- i)))
          (list
            [1 0] [-1 0]
            [0 1] [0 -1]
            [1 1] [1 -1]
            [-1 1] [-1 -1]
          ))))
))

;; Build the 2D cutout
(define cutout (recurse [0 0] 2/3 2))

;; Then reflect and cut it from a cube on all axes
(difference (box [-2 -2 -2] [2 2 2])
    cutout
    (reflect-yz cutout)
    (reflect-xz cutout))
```

So yeah, the 2d version is built up based on positioning beside/above and the 3d version is about cutouts.

I think I just have to change my mindset and this may work pretty well.

**Comment by lf94** [comment]
> The basic thing I don't see in your documentation is whether I can get a shape to return information about its size and position after it is created, to easily make a function that would, say, put one image "above" another.

In other code CAD I've created a "builder" object that tracks all this (basically a tree of shapes), but I've found it's mostly sufficient to just do all the dimensional and positioning calculations before the "rendering" work, i.e. `const rectDimensions = [w, h]; const rectPosition = [x,y]; rectangle(rectDimensions).move(rectPosition)` and so on.
